### PR TITLE
 Add __traits(isDynamicArray, T), second try

### PR DESCRIPTION
**Pull request by nordlow**
Discussion at https://forum.dlang.org/post/nhnofvzmijywsgqyyibw@forum.dlang.org.

Reopening of https://github.com/dlang/dmd/pull/9014.

This benchmark clearly shows the overhead of the templated approach for the case when trait argument is not an enum. In the enum case the overheads is gonna be greater because `isDynamicArray` is gonna instantiate twice for each enum parameter.

```D
// version = useBuiltin;           ///< Use new builtin trait __traits(isDynamicArray, ...)

private static alias ScalarTypes = AliasSeq!(bool,
                                             char, wchar, dchar,
                                             byte, ubyte,
                                             short, ushort,
                                             int, uint,
                                             long, ulong,
                                             float, double, real,
                                             cfloat, cdouble, creal,
                                             ifloat, idouble, ireal,
                                             string, wstring, dstring);

private static enum qualifiers = AliasSeq!("", "const", "inout", "immutable", "shared", "shared const");

static private template isDynamicArray(T)
{
    static if (is(T == U[], U))
        enum bool isDynamicArray = true;
    else static if (is(T U == enum))
        enum bool isDynamicArray = isDynamicArray!U;
    else
        enum bool isDynamicArray = false;
}

static foreach (T; ScalarTypes)
{
    static foreach (U; ScalarTypes)
    {
        static foreach (V; ScalarTypes)
        {
            mixin("struct ",
                  T, "_" ,U, "_" ,V,
                  " {",
                  T, " t; ",
                  U, " u; ",
                  V, " v; ",
                  "}");
            static foreach (qualifier; qualifiers)
            {
                version(useBuiltin)
                    static assert(__traits(isDynamicArray, mixin(qualifier, "(", T, "_", U, "_", V, ")")[])); // min over 10 runs: 2.62s.
                else
                    // static assert(is(mixin(qualifier, "(", T, "_", U, "_", V, ")")[] == X[], X)); // min over 10 runs: 2.75s
                    static assert(isDynamicArray!(mixin(qualifier, "(", T, "_", U, "_", V, ")")[])); // min over 10 runs: 3.19s
            }
        }
    }
}
```

There's also the issue of API consistency. If `isStaticArray` is present as a builtin trait a developer is gonna ask why isn't `isDynamicArray` present aswell?

### CODE REVIEW

File: changelog/trait_isdynamicarray.dd [hunk 1]
```diff
@@ -0,0 +1,9 @@
+Added new builtin trait `__traits(isDynamicArray, T)`.
+
+A new trait
+
+---
+__traits(isDynamicArray, T)
```
Comment (NilsLankila) [inline changelog/trait_isdynamicarray.dd:6]: The example should be comething that compiles.

File: changelog/trait_isdynamicarray.dd [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+---
+
+has been added that checks whether a type `T` is a dynamic array (slice) or not.
```
Comment (NilsLankila) [inline changelog/trait_isdynamicarray.dd:9]: slice between parens is confusing, I'd remove it.

**Review by ghost — APPROVED** [review]
The changelog entry needs a bit of work but otherwise this new trait is an easy win.

**Review by MoonlightSentinel — CHANGES_REQUESTED** [review]
The `enum` issue needs to be resolved and both `is` and this trait need to yield the same result

**Comment by wilzbach** [comment]
Instead of specialized benchmarks, why not replace the Phobos template with this trait and check for a difference?
(Should be easy to do as it's only changing the implementation in std.traits.)

**Comment by nordlow** [comment]
> Instead of specialized benchmarks, why not replace the Phobos template with this trait and check for a difference?
> (Should be easy to do as it's only changing the implementation in std.traits.)

These are micro-optimizations that might have an effect in very specific cases. It will most likely not have a measureable effect on the overload phobos compilation time.

**Comment by wilzbach** [comment]
> > Instead of specialized benchmarks, why not replace the Phobos template with this trait and check for a difference?
> > (Should be easy to do as it's only changing the implementation in std.traits.)
> 
> These are micro-optimizations that might have an effect in very specific cases as dmd, druntime and phobos gets further optimizations. It will most likely not have a measurable effect on the overall Phobos compilation time.

FWIW If you can't measure the impact on Phobos, it typically doesn't get merged here.

Also @UplinkCoder conflated internal changes in the compiler which can always be merged as they don't affect the user with adding a new public API which should have a measurable impact.

**Comment by andralex** [comment]
The measurements were ignored and this has become a sheer negotiation. Do we really want to do compiler development by horse-trading?

We should close this PR and move on. The person-hours we already spent on this are already overwhelming the cumulative future savings in compilation times from now through the Sun becoming a white dwarf.

**Comment by nordlow** [comment]
Ok with me to close this.

**Comment by RazvanN7** [comment]
@nordlow I'm taking the liberty to close this as it seems that a consensus has been reached.

**Pull request closed without merging**